src/common/secret/src/secret_manager.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap};
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+use anyhow::{anyhow, Context};
+use parking_lot::RwLock;
+use prost::Message;
+use risingwave_pb::catalog::PbSecret;
+use risingwave_pb::secret::secret_ref::RefAsType;
+use risingwave_pb::secret::PbSecretRef;
+use thiserror_ext::AsReport;
+
+use super::error::{SecretError, SecretResult};
+use super::SecretId;
+
+static INSTANCE: std::sync::OnceLock<LocalSecretManager> = std::sync::OnceLock::new();
+
+#[derive(Debug)]
+pub struct LocalSecretManager {
+    secrets: RwLock<HashMap<SecretId, Vec<u8>>>,
+    /// The local directory used to write secrets into file, so that it can be passed into some libararies
+    secret_file_dir: PathBuf,
+}
+
+impl LocalSecretManager {
+    /// Initialize the secret manager with the given temp file path, cluster id, and encryption key.
+    /// # Panics
+    /// Panics if fail to create the secret file directory.
+    pub fn init(temp_file_dir: Option<String>, cluster_id: String) {
```

Comment:
(Continue from https://github.com/risingwavelabs/risingwave/pull/17456#issuecomment-2207005580)

Introducing the `cluster_id` just for simulation tests is unacceptable to me. :(

> Maybe hack in risingwave_simulation::start to add a suffix?

It's not a hack. IIUC, simulation test runs multiple RW instance at the same time, so it should prepare separated working folders for them.

Reply:
Users can 1. start multiple RW clusters on 1 machine. 2. start multiple nodes on 1 machine.

Making users config a distinct path for each cluster and each node is not ideal. Misconfiging this can cause unexpected race on 1 directory. So not just cluster id, I actually think we should also add node id.

Reply:
> Making users config a distinct path for each cluster and each node is not ideal.

Hmmm, I think this is quite natural. If you are running multiple MySQL/PG service in one machine, you can't use the default data dir `/var/mysql/`.

Reply:
Then we need a way like lock file to make sure a dir won't be shared by two RW instance

Not only different RW instances, different working nodes and their replication can be potentially deployed on a single machine. 

Using RW cluster id to distinguish different RW instances can be used in other circumstances like connection name when two RW instance connection to the same up/downstream. So I guess users shouldn't be surprised to find a dir with uuid name

Reply:
Having the same question. Shall we accept the hypothesis that tempfile's name seldom collide?

---

In my design, a secret file on CN is never reused so we'd pass the content in each request. 
Using a local manager to keep the files live long enough is okay, but why we need to take the local one as the ground truth.

Reply:
> in each request
If you mean the request from meta to cn, we have a lot of those request including altering/create/recovery... also frontend need secret files.

> seldom collide
It can still happen. Do we have any reason not to handle this case?